File: aeron-client/src/test/cpp/concurrent/TermGapScannerTest.cpp
/*
 * Copyright 2014-2018 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


#include <gtest/gtest.h>

#include <thread>
#include "MockAtomicBuffer.h"
#include <concurrent/logbuffer/TermGapScanner.h>
#include <concurrent/logbuffer/LogBufferDescriptor.h>

using namespace aeron::concurrent::logbuffer;
using namespace aeron::concurrent::mock;
using namespace aeron::concurrent;
using namespace aeron::util;
using namespace aeron;

#define LOG_BUFFER_CAPACITY (LogBufferDescriptor::TERM_MIN_LENGTH)
#define TERM_ID 1

typedef std::array<std::uint8_t, LOG_BUFFER_CAPACITY> log_buffer_t;

class TermGapScannerTest : public testing::Test
{
public:
    TermGapScannerTest() :
        m_termBuffer(&m_logBuffer[0], m_logBuffer.size())
    {
        m_logBuffer.fill(0);
    }

    virtual void SetUp()
    {
        m_logBuffer.fill(0);
    }

protected:
    AERON_DECL_ALIGNED(log_buffer_t m_logBuffer, 16);
    MockAtomicBuffer m_termBuffer;
};

TEST_F(TermGapScannerTest, shouldReportGapAtBeginningOfBuffer)
{
    bool called = false;
    const util::index_t frameOffset = BitUtil::align(DataFrameHeader::LENGTH * 3, FrameDescriptor::FRAME_ALIGNMENT);
    const util::index_t highWaterMark = frameOffset + FrameDescriptor::ALIGNED_HEADER_LENGTH;

    EXPECT_CALL(m_termBuffer, getInt32Volatile(testing::_))
        .WillRepeatedly(testing::Return(0));
    EXPECT_CALL(m_termBuffer, getInt32Volatile(frameOffset))
        .WillOnce(testing::Return(DataFrameHeader::LENGTH));

    auto f = [&] (std::int32_t termId, AtomicBuffer& buffer, std::int32_t offset, std::int32_t length)
    {
        EXPECT_EQ(TERM_ID, termId);
        EXPECT_EQ(0, offset);
        EXPECT_EQ(frameOffset, length);
        called = true;
    };

    EXPECT_EQ(0, TermGapScanner::scanForGap(m_termBuffer, TERM_ID, 0, highWaterMark, f));

    EXPECT_TRUE(called);
}

TEST_F(TermGapScannerTest, shouldReportSingleGapWhenBufferNotFull)
{
    bool called = false;
    const std::int32_t tail = BitUtil::align(DataFrameHeader::LENGTH, FrameDescriptor::FRAME_ALIGNMENT);
    const std::int32_t highWaterMark = FrameDescriptor::FRAME_ALIGNMENT * 3;

    EXPECT_CALL(m_termBuffer, getInt32Volatile(testing::_)).WillRepeatedly(testing::Return(0));
    EXPECT_CALL(m_termBuffer, getInt32Volatile(tail - FrameDescriptor::ALIGNED_HEADER_LENGTH))
        .WillRepeatedly(testing::Return(DataFrameHeader::LENGTH));
    EXPECT_CALL(m_termBuffer, getInt32Volatile(tail))
        .WillRepeatedly(testing::Return(0));
    EXPECT_CALL(m_termBuffer, getInt32Volatile(highWaterMark - FrameDescriptor::ALIGNED_HEADER_LENGTH))
        .WillRepeatedly(testing::Return(DataFrameHeader::LENGTH));

    auto f = [&] (std::int32_t termId, AtomicBuffer& buffer, std::int32_t offset, std::int32_t length)
    {
        EXPECT_EQ(TERM_ID, termId);
        EXPECT_EQ(tail, offset);
        EXPECT_EQ(FrameDescriptor::ALIGNED_HEADER_LENGTH, length);
        called = true;
    };

    EXPECT_EQ(tail, TermGapScanner::scanForGap(m_termBuffer, TERM_ID, 0, highWaterMark, f));

    EXPECT_TRUE(called);
}

TEST_F(TermGapScannerTest, shouldReportSingleGapWhenBufferIsFull)
{
    bool called = false;
    const std::int32_t tail = LOG_BUFFER_CAPACITY - (FrameDescriptor::ALIGNED_HEADER_LENGTH * 2);
    const std::int32_t highWaterMark = LOG_BUFFER_CAPACITY;

    EXPECT_CALL(m_termBuffer, getInt32Volatile(testing::_)).WillRepeatedly(testing::Return(0));

    EXPECT_CALL(m_termBuffer, getInt32Volatile(tail - FrameDescriptor::ALIGNED_HEADER_LENGTH))
        .WillRepeatedly(testing::Return(DataFrameHeader::LENGTH));
    EXPECT_CALL(m_termBuffer, getInt32Volatile(tail))
        .WillRepeatedly(testing::Return(0));
    EXPECT_CALL(m_termBuffer, getInt32Volatile(highWaterMark - FrameDescriptor::ALIGNED_HEADER_LENGTH))
        .WillRepeatedly(testing::Return(DataFrameHeader::LENGTH));

    auto f = [&] (std::int32_t termId, AtomicBuffer& buffer, std::int32_t offset, std::int32_t length)
    {
        EXPECT_EQ(TERM_ID, termId);
        EXPECT_EQ(tail, offset);
        EXPECT_EQ(FrameDescriptor::ALIGNED_HEADER_LENGTH, length);
        called = true;
    };

    EXPECT_EQ(tail, TermGapScanner::scanForGap(m_termBuffer, TERM_ID, tail, highWaterMark, f));

    EXPECT_TRUE(called);
}

TEST_F(TermGapScannerTest, shouldReportNoGapWhenHwmIsInPadding)
{
    bool called = false;
    const std::int32_t paddingLength = FrameDescriptor::ALIGNED_HEADER_LENGTH * 2;
    const std::int32_t tail = LOG_BUFFER_CAPACITY - paddingLength;
    const std::int32_t highWaterMark = LOG_BUFFER_CAPACITY - paddingLength + DataFrameHeader::LENGTH;

    EXPECT_CALL(m_termBuffer, getInt32Volatile(tail))
        .WillRepeatedly(testing::Return(paddingLength));
    EXPECT_CALL(m_termBuffer, getInt32Volatile(tail + DataFrameHeader::LENGTH))
        .WillRepeatedly(testing::Return(0));

    auto f = [&] (std::int32_t termId, AtomicBuffer& buffer, std::int32_t offset, std::int32_t length)
    {
        called = true;
    };

    EXPECT_EQ(LOG_BUFFER_CAPACITY, TermGapScanner::scanForGap(m_termBuffer, TERM_ID, tail, highWaterMark, f));

    EXPECT_FALSE(called);
}

File: aeron-client/src/test/java/io/aeron/logbuffer/TermGapScannerTest.java
/*
 * Copyright 2014-2025 Real Logic Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.aeron.logbuffer;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import io.aeron.protocol.DataHeaderFlyweight;
import org.agrona.concurrent.UnsafeBuffer;

import static io.aeron.logbuffer.FrameDescriptor.FRAME_ALIGNMENT;
import static org.agrona.BitUtil.align;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;
import static org.mockito.Mockito.verifyNoMoreInteractions;

class TermGapScannerTest
{
    private static final int LOG_BUFFER_CAPACITY = LogBufferDescriptor.TERM_MIN_LENGTH;
    private static final int TERM_ID = 1;
    private static final int HEADER_LENGTH = DataHeaderFlyweight.HEADER_LENGTH;

    private final UnsafeBuffer termBuffer = mock(UnsafeBuffer.class);
    private final TermGapScanner.GapHandler gapHandler = mock(TermGapScanner.GapHandler.class);

    @BeforeEach
    void setUp()
    {
        when(termBuffer.capacity()).thenReturn(LOG_BUFFER_CAPACITY);
    }

    @Test
    void shouldReportGapAtBeginningOfBuffer()
    {
        final int frameOffset = align(HEADER_LENGTH * 3, FRAME_ALIGNMENT);
        final int highWaterMark = frameOffset + align(HEADER_LENGTH, FRAME_ALIGNMENT);

        when(termBuffer.getIntVolatile(frameOffset)).thenReturn(HEADER_LENGTH);

        assertEquals(0, TermGapScanner.scanForGap(termBuffer, TERM_ID, 0, highWaterMark, gapHandler));

        verify(gapHandler).onGap(TERM_ID, 0, frameOffset);
        verifyNoMoreInteractions(gapHandler);
    }

    @Test
    void shouldReportSingleGapWhenBufferNotFull()
    {
        final int tail = align(HEADER_LENGTH, FRAME_ALIGNMENT);
        final int highWaterMark = FRAME_ALIGNMENT * 3;

        when(termBuffer.getIntVolatile(tail - align(HEADER_LENGTH, FRAME_ALIGNMENT))).thenReturn(HEADER_LENGTH);
        when(termBuffer.getIntVolatile(tail)).thenReturn(0);
        when(termBuffer.getIntVolatile(highWaterMark - align(HEADER_LENGTH, FRAME_ALIGNMENT)))
            .thenReturn(HEADER_LENGTH);

        assertEquals(tail, TermGapScanner.scanForGap(termBuffer, TERM_ID, tail, highWaterMark, gapHandler));

        verify(gapHandler).onGap(TERM_ID, tail, align(HEADER_LENGTH, FRAME_ALIGNMENT));
        verifyNoMoreInteractions(gapHandler);
    }

    @Test
    void shouldReportSingleGapWhenBufferIsFull()
    {
        final int tail = LOG_BUFFER_CAPACITY - (align(HEADER_LENGTH, FRAME_ALIGNMENT) * 2);
        final int highWaterMark = LOG_BUFFER_CAPACITY;

        when(termBuffer.getIntVolatile(tail - align(HEADER_LENGTH, FRAME_ALIGNMENT))).thenReturn(HEADER_LENGTH);
        when(termBuffer.getIntVolatile(tail)).thenReturn(0);
        when(termBuffer.getIntVolatile(highWaterMark - align(HEADER_LENGTH, FRAME_ALIGNMENT)))
            .thenReturn(HEADER_LENGTH);

        assertEquals(tail, TermGapScanner.scanForGap(termBuffer, TERM_ID, tail, highWaterMark, gapHandler));

        verify(gapHandler).onGap(TERM_ID, tail, align(HEADER_LENGTH, FRAME_ALIGNMENT));
        verifyNoMoreInteractions(gapHandler);
    }

    @Test
    void shouldReportNoGapWhenHwmIsInPadding()
    {
        final int paddingLength = align(HEADER_LENGTH, FRAME_ALIGNMENT) * 2;
        final int tail = LOG_BUFFER_CAPACITY - paddingLength;
        final int highWaterMark = LOG_BUFFER_CAPACITY - paddingLength + HEADER_LENGTH;

        when(termBuffer.getIntVolatile(tail)).thenReturn(paddingLength);
        when(termBuffer.getIntVolatile(tail + HEADER_LENGTH)).thenReturn(0);

        assertEquals(
            LOG_BUFFER_CAPACITY, TermGapScanner.scanForGap(termBuffer, TERM_ID, tail, highWaterMark, gapHandler));

        verifyNoInteractions(gapHandler);
    }

    @Test
    void shouldReportSingleHeaderGap()
    {
        final int offset = 8192 + 384;
        when(termBuffer.getIntVolatile(offset)).thenReturn(0);
        when(termBuffer.getIntVolatile(offset + HEADER_LENGTH)).thenReturn(128);

        assertEquals(
            offset, TermGapScanner.scanForGap(termBuffer, TERM_ID, offset, LOG_BUFFER_CAPACITY, gapHandler));

        verify(termBuffer).getIntVolatile(offset);
        verify(termBuffer).getIntVolatile(offset + HEADER_LENGTH);
        verify(gapHandler).onGap(TERM_ID, offset, HEADER_LENGTH);
        verifyNoMoreInteractions(gapHandler, termBuffer);
    }

    @Test
    void shouldReportGapAtTheEndOfTheBuffer()
    {
        final int offset = LOG_BUFFER_CAPACITY - 128;
        when(termBuffer.getIntVolatile(offset)).thenReturn(0);

        assertEquals(
            offset, TermGapScanner.scanForGap(termBuffer, TERM_ID, offset, LOG_BUFFER_CAPACITY, gapHandler));

        verify(termBuffer).getIntVolatile(offset);
        verify(termBuffer).getIntVolatile(offset + HEADER_LENGTH);
        verify(termBuffer).getIntVolatile(offset + 2 * HEADER_LENGTH);
        verify(termBuffer).getIntVolatile(offset + 3 * HEADER_LENGTH);
        verify(gapHandler).onGap(TERM_ID, offset, 128);
        verifyNoMoreInteractions(gapHandler, termBuffer);
    }
}
